test/infrastructure/vcsim/controllers/envsubst_controller.go

```diff
@@ -0,0 +1,252 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"crypto/rand"
+	"crypto/rsa"
+	"fmt"
+	"strconv"
+	"strings"
+	"sync"
+
+	"github.com/pkg/errors"
+	"golang.org/x/crypto/ssh"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/klog/v2"
+	"k8s.io/utils/ptr"
+	"sigs.k8s.io/cluster-api/util/patch"
+	"sigs.k8s.io/cluster-api/util/predicates"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+
+	vcsimv1 "sigs.k8s.io/cluster-api-provider-vsphere/test/infrastructure/vcsim/api/v1alpha1"
+)
+
+type EnvSubstReconciler struct {
+	Client         client.Client
+	SupervisorMode bool
+
+	PodIP   string
+	sshKeys map[string]string
+	lock    sync.RWMutex
+
+	// WatchFilterValue is the label value used to filter events prior to reconciliation.
+	WatchFilterValue string
+}
+
+// +kubebuilder:rbac:groups=vcsim.infrastructure.cluster.x-k8s.io,resources=envsubsts,verbs=get;list;watch;patch
+// +kubebuilder:rbac:groups=vcsim.infrastructure.cluster.x-k8s.io,resources=envsubsts/status,verbs=get;update;patch
+
+func (r *EnvSubstReconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Result, reterr error) {
+	log := ctrl.LoggerFrom(ctx)
+
+	// Fetch the EnvSubst instance
+	envSubst := &vcsimv1.EnvSubst{}
+	if err := r.Client.Get(ctx, req.NamespacedName, envSubst); err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+
+	// Fetch the VCenter instance
+	if envSubst.Spec.VCenter == "" {
+		return ctrl.Result{}, errors.New("Spec.VCenter cannot be empty")
+	}
+
+	vCenter := &vcsimv1.VCenter{}
+	if err := r.Client.Get(ctx, client.ObjectKey{
+		Namespace: envSubst.Namespace,
+		Name:      envSubst.Spec.VCenter,
+	}, vCenter); err != nil {
+		return ctrl.Result{}, errors.Wrapf(err, "failed to get VCenter")
+	}
+	log = log.WithValues("VCenter", klog.KObj(vCenter))
```

Comment:
```suggestion
	log = log.WithValues("VCenter", klog.KObj(vCenter))
	ctx = ctrl.LoggerInto(ctx, log)
```

Let's directly update the ctx. So the logger is already up-to-date wherever we use the ctx